### PR TITLE
Feat/hierarchical workflow tracing

### PR DESCRIPTION
**Pull request by serglom21**
## Summary

This PR transforms the GitHub Actions workflow tracing from individual job transactions to a single workflow transaction containing nested job and step spans, providing better visibility into workflow timing and structure.

## Problem that I am attempting to solve

**Before**: Individual job transactions created a flat structure in Sentry, making it difficult to understand workflow timing and relationships.
**After**: Single workflow transaction with proper hierarchical spans showing:
- Workflow-level timing and status
- Individual job spans as children
- Step spans as children of their respective jobs

## Key Changes

**Core Implementation**:
- WorkflowTracer: New class for creating single workflow transactions with nested spans
- WorkflowJobCollector: Collects jobs from workflow runs and sends workflow-level traces
- Enhanced WebAppHandler: Integrates job collection and workflow tracing

## Trace Structure:

```
workflow: Multi-Job Test (transaction)
├─ security-scan (job span)
│  ├─ Set up job (step span)
│  ├─ Checkout code (step span)
│  └─ Run security scan (step span)
├─ performance-tests (job span)
│  ├─ Set up job (step span)
│  ├─ Setup performance test environment (step span)
│  └─ Run performance tests (step span)
└─ backend-tests (job span)
   ├─ Set up job (step span)
   ├─ Setup Python (step span)
   └─ Run backend unit tests (step span)
```

## Files Changed

- `src/workflow_tracer.py` - New WorkflowTracer implementation
- `src/web_app_handler.py` - Enhanced with WorkflowJobCollector
- `src/enhanced_web_app_handler.py` - Alternative handler implementation

## Before

Screenshot 2025-10-01 at 7 13 42 AM

## After

Screenshot 2025-10-01 at 7 14 45 AM

### CODE REVIEW

**Comment by armenzg** [comment]
bugbot run

File: src/web_app_handler.py [hunk 1]
```diff
@@ -4,21 +4,240 @@
...
-from src.sentry_config import fetch_dsn_for_github_org
+from .workflow_tracer import WorkflowTracer
+from .sentry_config import fetch_dsn_for_github_org
 
 LOGGING_LEVEL = os.environ.get("LOGGING_LEVEL", logging.INFO)
 logger = logging.getLogger(__name__)
 logger.setLevel(LOGGING_LEVEL)
 
 
+class WorkflowJobCollector:
```
Comment (armenzg) [inline src/web_app_handler.py]: Move this into a new module and import it here.

File: src/web_app_handler.py [hunk 2]
```diff
@@ -4,21 +4,240 @@
...
+            
+            logger.info(f"Added job {job['name']} (ID: {job_id}) to workflow run {run_id}")
+            
+            # Smart workflow completion detection
+            jobs_count = len(self.workflow_jobs[run_id])
+            if run_id not in self.processed_workflows:
+                if self._should_process_workflow(run_id, jobs_count):
+                    logger.info(f"Workflow run {run_id} has {jobs_count} jobs, setting timer to process in 2 seconds")
+                    # Set a short timer to allow all jobs to arrive
+                    timer = threading.Timer(2.0, self._process_workflow_immediately, args=[run_id])
```
Comment (armenzg) [inline src/web_app_handler.py]: No magic numbers please: https://refactoring.guru/replace-magic-number-with-symbolic-constant

File: src/web_app_handler.py [hunk 3]
```diff
@@ -4,21 +4,240 @@
...
+                    self._send_workflow_trace(run_id)
+                else:
+                    logger.info(f"Not all jobs complete for workflow run {run_id}, skipping")
+        except Exception as e:
+            logger.error(f"Error processing workflow run {run_id} immediately: {e}", exc_info=True)
+            # Ensure cleanup happens even if there's an exception
+            self._cleanup_workflow_run(run_id)
+    
+    def _process_workflow_delayed(self, run_id: int):
+        """Process workflow after delay to allow all jobs to arrive"""
```
Comment (armenzg) [inline src/web_app_handler.py]: What do we do when the jobs do not arrive? or arrive hours later?

**Review by armenzg — CHANGES_REQUESTED** [review]
Thanks Sergio for the contribution but there are several issues that need addressing before merging this.

The main one is that there are no tests and there are various principles of clean code are not being followed.

You can use Cursor to refactor the PR using "clean code principles".

Also make changes so we can handle the old style and the new style while we verify the changes are working. We may want to only do this for the Sentry org events. We would also need to verify that current CI dashboards in the getsentry repo would not break.

File: src/web_app_handler.py (hunk 1 above)
Comment (serglom21) [reply src/web_app_handler.py]: done!

File: src/web_app_handler.py (hunk 3 above)
Comment (serglom21) [reply src/web_app_handler.py]: good q: I implemented a GitHub API verification to handle jobs that arrive after a workflow is processed. I basically call the API to check the workflow status

**Comment by serglom21** [comment]
Thanks @armenzg! I have implemented the changes above and added some tests for the new functionality and edge cases. More details:

- Extracted WorkflowJobCollector into its own module (src/workflow_job_collector.py)
- Replaced magic numbers with constants

I added some env variable handling to:
  - ENABLE_HIERARCHICAL_TRACING to toggle new vs legacy tracing
  - SENTRY_ORG_ONLY to restrict to Sentry org during testing
  - Legacy individual job tracing remains available as fallback

- Implemented clean code principles. Could you please review again? Thanks
